File: C1.UWP.FlexGrid/CS/FlexGrid101/CellFactory/MusicCellFactory.cs
using System;
using System.Net;
using System.Windows;
using System.Collections.Generic;
using System.Linq;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Animation;
using C1.Xaml.FlexGrid;

namespace FlexGrid101
{
    // cell factory used to create media cells
    public class MusicCellFactory : CellFactory
    {
        static Thickness _emptyThickness = new Thickness(0);
        public List<RatingCell> _ratings = new List<RatingCell>();

        // bind a cell to a range
        public override void CreateCellContent(C1FlexGrid grid, Border bdr, CellRange range)
        {
            // get row/col
            var row = grid.Rows[range.Row];
            var col = grid.Columns[range.Column];
            var gr = row as GroupRow;

            // no border for group rows
            if (gr != null)
            {
                bdr.BorderThickness = _emptyThickness;
            }

            // bind first cell in each group row to collapse/expand
            if (gr != null && range.Column == 0)
            {
                BindGroupRowCell(grid, bdr, range);
                return;
            }

            // bind regular data cells
            switch (col.ColumnName)
            {
                // show names with images
                case "Name":
                    bdr.Child = new SongCell(row);
                    return;

                // show ratings with stars
                case "Rating":
                    var song = row.DataItem as Song;
                    if (song != null && gr == null)
                    {
                        var cell = new RatingCell();
                        cell.SetBinding(RatingCell.RatingProperty, col.Binding);
                        bdr.Child = cell;
                    }
                    return;
            }

            // default binding
            base.CreateCellContent(grid, bdr, range);
        }

        // bind a cell to a group row
        void BindGroupRowCell(C1FlexGrid grid, Border bdr, CellRange range)
        {
            // get row, group row
            var row = grid.Rows[range.Row];
            var gr = row as GroupRow;

            // show group caption/image on first column
            if (range.Column == 0)
            {
                // build a custom data item if necessary
                if (gr.DataItem == null)
                {
                    gr.DataItem = BuildGroupDataItem(gr);
                }

                // get type of cell we need
                Type cellType = gr.Level == 0 ? typeof(ArtistCell) : typeof(AlbumCell);

                // create the cell
                bdr.Child = gr.Level == 0 
                    ? (UIElement)new ArtistCell(row)
                    : (UIElement)new AlbumCell(row);
            }
        }

        // build a song to represent a group
        // the GetChildDataItems method returns all songs that belong
        // to this node, and the LINQ statement below calculates the total 
        // size, length, and average rating for the album/artist.
        Song BuildGroupDataItem(GroupRow gr)
        {
            var gs = gr.GetDataItems().OfType<Song>();
            return new Song()
                {
                    Name = gr.Group.Group.ToString(),
                    Size = (long)gs.Sum(s => s.Size),
                    Duration = (long)gs.Sum(s => s.Duration),
                    Rating = (int)(gs.Average(s => s.Rating) + 0.5)
                };
        }
    }
}

File: C1.UWP.FlexGrid/VB/FlexGridSamples/CellFactory/MusicCellFactory.vb
Imports C1.Xaml.FlexGrid
Imports Windows.UI.Xaml.Media.Animation
Imports Windows.UI.Xaml.Media
Imports Windows.UI.Xaml.Controls
Imports Windows.UI.Xaml
Imports System.Linq
Imports System.Collections.Generic
Imports System.Windows
Imports System.Net
Imports System

' cell factory used to create media cells
Public Class MusicCellFactory
    Inherits CellFactory
    Shared _emptyThickness As New Thickness(0)
    Public _ratings As New List(Of RatingCell)()

    ' bind a cell to a range
    Public Overrides Sub CreateCellContent(grid As C1FlexGrid, bdr As Border, range As CellRange)
        ' get row/col
        Dim row As Row = grid.Rows(range.Row)
        Dim col As Column = grid.Columns(range.Column)
        Dim gr As GroupRow = TryCast(row, GroupRow)

        ' no border for group rows
        If gr IsNot Nothing Then
            bdr.BorderThickness = _emptyThickness
        End If

        ' bind first cell in each group row to collapse/expand
        If gr IsNot Nothing AndAlso range.Column = 0 Then
            BindGroupRowCell(grid, bdr, range)
            Return
        End If

        ' bind regular data cells
        Select Case col.ColumnName
                ' show names with images
            Case "Name"
                bdr.Child = New SongCell(row)
                Return

                ' show ratings with stars
            Case "Rating"
                Dim song As Song = TryCast(row.DataItem, Song)
                If song IsNot Nothing AndAlso gr Is Nothing Then
                    Dim cell As New RatingCell()
                    cell.SetBinding(RatingCell.RatingProperty, col.Binding)
                    bdr.Child = cell
                End If
                Return
        End Select

        ' default binding
        MyBase.CreateCellContent(grid, bdr, range)
    End Sub

    ' bind a cell to a group row
    Sub BindGroupRowCell(grid As C1FlexGrid, bdr As Border, range As CellRange)
        ' get row, group row
        Dim row As Row = grid.Rows(range.Row)
        Dim gr As GroupRow = TryCast(row, GroupRow)

        ' show group caption/image on first column
        If range.Column = 0 Then
            ' build a custom data item if necessary
            If gr.DataItem Is Nothing Then
                gr.DataItem = BuildGroupDataItem(gr)
            End If

            ' get type of cell we need
            Dim cellType As Type = If(gr.Level = 0, GetType(ArtistCell), GetType(AlbumCell))

            ' create the cell
            bdr.Child = If(gr.Level = 0, CType(New ArtistCell(row), UIElement), CType(New AlbumCell(row), UIElement))
        End If
    End Sub

    ' build a song to represent a group
    ' the GetChildDataItems method returns all songs that belong
    ' to this node, and the LINQ statement below calculates the total 
    ' size, length, and average rating for the album/artist.
    Function BuildGroupDataItem(gr As GroupRow) As Song
        Dim gs As IEnumerable(Of Song) = gr.GetDataItems().OfType(Of Song)()
        Return New Song() With {
            .Name = gr.Group.Group.ToString(),
            .Size = CType(gs.Sum(Function(s) s.Size), Long),
            .Duration = CType(gs.Sum(Function(s) s.Duration), Long),
            .Rating = CType((gs.Average(Function(s) s.Rating) + 0.5), Integer)
        }
    End Function
End Class